gateway-service/src/main/java/org/zowe/apiml/gateway/security/mapping/model/MapperResponse.java

```diff
@@ -28,5 +35,40 @@ public class MapperResponse {
     @JsonProperty("racfReturnCode")
     private int racfRc;
     @JsonProperty("racfReasonCode")
-    private int reasonCode;
+    private int racfRs;
+
+    @InjectApimlLogger
+    private final ApimlLogger apimlLog = ApimlLogger.empty();
+
+    public String toString() {
+        return "User: " + userId + ", rc=" + rc + ", safRc=" + safRc + ", racfRc=" + racfRc + ", racfRs=" + racfRs;
+    }
+
+    public boolean isOIDCResultValid() {
+        // Some codes may be 4 and the result is still valid. But deny unless we know it for sure
+        if (rc == 0 || safRc == 0 || racfRc == 0 || racfRs == 0) {
+            return true;
+        }
+
+        if (rc == 8 && safRc == 8 && racfRc == 8) {
+            if (racfRs == 44) {
+                apimlLog.log(OIDC_FAILED_MESSAGE_KEY,
+                    "The Registry Name or supplied distributed identity is all blanks (x'20'), all nulls" +
+                        " (x'00'), or a combination of blanks and nulls.");
+
+                return false;
+            }
+            if (racfRs == 48) {
+                apimlLog.log(OIDC_FAILED_MESSAGE_KEY,
+                    "There is no distributed identity filter mapping the supplied distributed identity to" +
+                        " a SAF user ID, or the IDIDMAP SAF general resource class is not active or not RACLISTed.");
+
+                return false;
+            }
+        }
+
+        apimlLog.log(OIDC_FAILED_MESSAGE_KEY, "SAF response: " + this);
```

Comment:
Why do we log this? Do we expect an unexpected outcome from the mapping?

Reply:
Current error combinations cover 90% of all errors. I guess it is sufficient for now.